lib/chunks.sql

```diff
@@ -190,27 +190,83 @@ WITH
         AND a.course_instance_id = ci.id
       )
   ),
-  chunks_to_delete AS (
+  -- We use separate CTEs for each chunk type so that Postgres doesn't tye to
+  -- compute a huge cross product of the chunks + metadata on `type` as an
+  -- intermediate step. This was O(n^2) with the number of chunks to delete.
+  question_chunks_to_delete AS (
     SELECT
       id
     FROM
       chunks
       JOIN chunks_metadata AS cm ON (
-        chunks.course_id = $course_id
-        AND chunks.type = cm.type
-        AND CASE
-          WHEN (
-            cm.type = 'elements'
-            OR cm.type = 'elementExtensions'
-            OR cm.type = 'clientFilesCourse'
-            OR cm.type = 'serverFilesCourse'
-          ) THEN TRUE
-          WHEN (cm.type = 'clientFilesCourseInstance') THEN cm.course_instance_id = chunks.course_instance_id
-          WHEN (cm.type = 'clientFilesAssessment') THEN cm.assessment_id = chunks.assessment_id
-          WHEN (cm.type = 'question') THEN cm.question_id = chunks.question_id
-          ELSE FALSE
-        END
+        cm.type = chunks.type
+        AND cm.question_id = chunks.question_id
+      )
+    WHERE
+      chunks.course_id = $course_id
+      AND chunks.type = 'question'
+  ),
+  client_files_course_instance_chunks_to_delete AS (
+    SELECT
+      id
+    FROM
+      chunks
+      JOIN chunks_metadata AS cm ON (
+        cm.type = chunks.type
+        AND cm.course_instance_id = chunks.course_instance_id
+      )
+    WHERE
+      chunks.course_id = $course_id
+      AND chunks.type = 'clientFilesCourseInstance'
+  ),
+  client_files_assessment_chunks_to_delete AS (
+    SELECT
+      id
+    FROM
+      chunks
+      JOIN chunks_metadata AS cm ON (
+        cm.type = chunks.type
+        AND cm.assessment_id = chunks.assessment_id
+      )
+    WHERE
+      chunks.course_id = $course_id
+      AND chunks.type = 'clientFilesAssessment'
+  ),
+  other_chunks_to_delete AS (
+    SELECT
+      id
+    FROM
+      chunks
+      JOIN chunks_metadata AS cm ON (cm.type = chunks.type)
+    WHERE
+      chunks.course_id = $course_id
+      AND (
+        cm.type = 'elements'
+        OR cm.type = 'elementExtensions'
+        OR cm.type = 'clientFilesCourse'
+        OR cm.type = 'serverFilesCourse'
       )
```

Comment:
```suggestion
      AND cm.type IN ('elements', 'elementExtensions', 'clientFilesCourse', 'serverFilesCourse')
```

`IN` has better use of indices than `OR`.

Reply:
Changed in 63c5341.

Reply:
This bad performance of OR is something I need to remind myself about. I need to keep remembering that excellent blog post that you found, @jonatanschroeder.